Add mcxc242 base enablement

michal-smola opened the pull request:
Add basic enablement of mcxc 242, 142 and 141 socs in device tree.
Add basic enablement of mcxc series.
Add basic enablement of frdm_mcxc242 board.
Fix build error caused by uart_mcux_lpuart.c driver

michal-smola committed:
soc: nxp mcxc: Add support for NXP MCXC series

Add initial suport for NXP MCXC series

Signed-off-by: Michal Smola <michal.smola@nxp.com>

michal-smola committed:
boards: frdm_mcxc242: Add basic support for board

Add basic support for NXP frdm_mcxc242 board.

Signed-off-by: Michal Smola <michal.smola@nxp.com>

michal-smola committed:
drivers: uart: Fix build error for lpuart without modem support

LPUART without modem support is used in some NXP socs. Build error
occurs for such socs because of accessing unavailable register.
Fix it by adding #if directive for checking modem support availability.

Signed-off-by: Michal Smola <michal.smola@nxp.com>

nordicjm commented on boards/nxp/frdm_mcxc242/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+# Copyright 2024 NXP
+#
+# SPDX-License-Identifier: Apache-2.0
+#
```
empty file can go

michal-smola replied:
the file was removed

nordicjm commented on boards/nxp/frdm_mcxc242/doc/frdm_mcxc242.jpg:
convert to wepb and put through https://tinypng.com/

michal-smola replied:
Done

nordicjm commented on soc/nxp/mcx/mcxc/Kconfig:
```diff
@@ -0,0 +1,148 @@
+
+if HAS_OSC
+
+choice MCXC_OSC_MODE
+	prompt "Oscillator Mode Selection"
+	default MCXC_OSC_EXTERNAL
+
+config MCXC_OSC_EXTERNAL
+	bool "External reference clock"
+	help
+	  Set this option to use the oscillator in external reference clock mode.
+
+config MCXC_OSC_LOW_POWER
+	bool "Low power oscillator"
+	help
+	  Set this option to use the oscillator in low-power mode.
+
+config MCXC_OSC_HIGH_GAIN
+	bool "High gain oscillator"
+	help
+	  Set this option to use the oscillator in high-gain mode.
```
move to dts

nordicjm replied:
fix this in the original commit not adding something wrong then changing to to be right in a later commit

nordicjm commented on soc/nxp/mcx/mcxc/Kconfig:
```diff
@@ -0,0 +1,148 @@
+
+endchoice
+
+config OSC_XTAL0_FREQ
+	int "External oscillator frequency"
+	help
+	  Set the external oscillator frequency in Hz. This should be set by the
+	  board's defconfig.
```
dts

nordicjm commented on soc/nxp/mcx/mcxc/Kconfig:
```diff
@@ -0,0 +1,148 @@
+
+if HAS_MCG
+
+config MCG_FCRDIV
+	int "Low-frequency internal reference clock divider 1"
+	range 0 7
+	default 0
+	help
+	  Selects the amount to divide down the low-frequency internal
+	  reference clock as exponent of power of 2.
+
+config MCG_LIRCDIV2
+	int "Low-frequency internal reference clock divider 2"
+	range 0 7
+	default 0
+	help
+	  Selects the amount to divide down the low-frequency internal
+	  reference clock as exponent of power of 2.
```
dts

nordicjm commented on soc/nxp/mcx/mcxc/Kconfig:
```diff
@@ -0,0 +1,148 @@
+
+if MCXC_FLASH_CONFIG
+
+config MCXC_FLASH_CONFIG_OFFSET
+	hex "MCXC flash configuration field offset"
+	default 0x400
+
+config MCXC_FLASH_CONFIG_FSEC
+	hex "Flash security byte (FSEC)"
+	range 0 0xff
+	default 0xfe
+	help
+	  Configures the reset value of the FSEC register, which includes
+	  backdoor key access, mass erase, factory access, and flash security
+	  options.
+
+config MCXC_FLASH_CONFIG_FOPT
+	hex "Flash nonvolatile option byte (FOPT)"
+	range 0 0xff
+	default 0xff
+	help
+	  Configures the reset value of the FOPT register, which includes boot,
+	  NMI, and EzPort options.
+
+config MCXC_FLASH_CONFIG_FEPROT
+	hex "EEPROM protection byte (FEPROT)"
+	range 0 0xff
+	default 0xff
+	help
+	  Configures the reset value of the FEPROT register for FlexNVM
+	  devices. For program flash only devices, this byte is reserved.
+
+config MCXC_FLASH_CONFIG_FDPROT
+	hex "Data flash protection byte (FDPROT)"
+	range 0 0xff
+	default 0xff
+	help
+	  Configures the reset value of the FDPROT register for FlexNVM
+	  devices. For program flash only devices, this byte is reserved.
+
```
dts for anything that is hardware etc.

michal-smola committed:
dts: bindings: nxp kinetis mcg: Add optional frequency dividers

fcrdiv and lircdiv2 dividers properties are not configurable in device
tree. Add the properties.

Signed-off-by: Michal Smola <michal.smola@nxp.com>

michal-smola committed:
dts: bindings: kinetis-ftfa: Add configuration properties

Add configuration settings for the FOPT, FSEC, and flash configuration
offset to the FTFA module binding.

Signed-off-by: Michal Smola <michal.smola@nxp.com>

dleach02 commented:
@michal-smola I've updated your PR to remove west.yml change now that we have sdk2.16 merged

decsny commented on dts/arm/nxp/nxp_mcxc_common.dtsi:
are the contents of this file actually going to be common to all MCXC down the line like the name implies?

michal-smola replied:
Yes, the plan is to have common file to all MCXC. There is about 90% or even greater overlap of IPs for MCXC SOCs.

decsny commented on include/zephyr/dt-bindings/clock/kinetis_sim.h:
```diff
@@ -15,6 +15,9 @@
 #define KINETIS_SIM_DMA_CLK     KINETIS_SIM_CORESYS_CLK
 #define KINETIS_SIM_SIM_SOPT7   7
+#define KINETIS_SIM_OSCERCLK    8
+#define KINETIS_SIM_MCGIRCLK    12
```
these changes seem to be in the wrong commit.

michal-smola replied:
It is intentional. The defines can be used to select clock source for LPUART in devicetree. Do you prefer to have this change in another commit?

danieldegrasse replied:
Adding here- please do move this change to another commit

michal-smola replied:
Moved to separate commit.

decsny commented on boards/nxp/frdm_mcxc242/CMakeLists.txt:
```diff
@@ -0,0 +1 @@
+# SPDX-License-Identifier: Apache-2.0
```
why add an empty file with only a license?

michal-smola replied:
Removed the file .

decsny commented on drivers/serial/uart_mcux_lpuart.c:
```diff
@@ -1092,13 +1092,17 @@ static int mcux_lpuart_configure_init(const struct device *dev, const struct uar
 	LPUART_Init(config->base, &uart_config, clock_freq);
 
+#if defined(FSL_FEATURE_LPUART_HAS_MODEM_SUPPORT) && \
+	FSL_FEATURE_LPUART_HAS_MODEM_SUPPORT
```
the ci check wants this to be on one line

michal-smola replied:
CI check does not seem to fail. There has been one more instance of the same line in the file before my change already. Unified style of the file would be nice. Can I leave the added lines as is, or do you prefer to change all occurrences to the one line format?

danieldegrasse replied:
This commit should come before the commits adding support for the MCXC242, if that is the SOC causing this build error

michal-smola replied:
Moved the commit.

decsny reviewed (commented):
reviewed the dts, board, and driver commits

michal-smola committed:
dts: bindings: Add binding for NXP mcxc oscillator

Devicetree binding for NXP mcxc oscillator is not available.
Add the binding to be able to configure the oscillator in devicetree.

Signed-off-by: Michal Smola <michal.smola@nxp.com>

michal-smola commented:
I've reworked MCG, FTFA and OSC configuration from Kconfig to Devicetree to address comments by @nordicjm. Could you review it please?

michal-smola commented:
I've reorganized the commits. Commits with dts bindings additions first, then the original commits fixed.

dleach02 commented:
@nordicjm can you verify if the change addressed your request?

nordicjm commented on dts/bindings/clock/nxp,mcxc-osc.yaml:
```diff
@@ -0,0 +1,28 @@
+properties:
+  load-capacitance:
+    type: int
+    enum: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
```
@gmarull should this be like this or just have the pf values directly?

gmarull commented on dts/bindings/clock/nxp,mcxc-osc.yaml:
```diff
@@ -0,0 +1,28 @@
+include: fixed-clock.yaml
+
+properties:
+  load-capacitance:
```
please, add units here

michal-smola replied:
The property is not load capacitance directly, but rather configuration of the load. Changed the misleading property name to load-configuration (without a unit). Meaning is explained in the description of the property. @gmarull please review.

gmarull replied:
Then this is against bindings good practices, use meaningful units.

michal-smola replied:
Changed to femtofarads.

michal-smola replied:
@gmarull could you review the changes, please?

dleach02 commented:
@gmarull can you check to see if your change request has been addressed?

gmarull commented on dts/bindings/flash_controller/nxp,kinetis-ftfa.yaml:
```diff
@@ -3,3 +3,22 @@ description: NXP Kinetis Flash Memory Module A (FTFA)
+  config-field-offset:
+    type: int
+    description: |
+      "Flash configuration field offset"
```
redundant `"`

michal-smola replied:
Removed redundant "

gmarull commented on dts/bindings/clock/nxp,mcxc-osc.yaml:
```diff
@@ -0,0 +1,40 @@
+include: fixed-clock.yaml
+
+properties:
+  load-capacitance-femtofarads:
+    type: int
+    enum:
+      - 0
+      - 2000
+      - 4000
+      - 6000
+      - 8000
+      - 10000
+      - 12000
+      - 14000
+      - 16000
+      - 18000
+      - 20000
+      - 22000
+      - 24000
+      - 26000
+      - 28000
+      - 30000
```
why are you using femto units when all values seem to be scaled by 1000?

michal-smola replied:
There is no apparent reason. I will change it to picofarads.

michal-smola replied:
Changed to picofarads. @gmarull could you review, please?

danieldegrasse commented on soc/nxp/mcx/mcxc/pinctrl_soc.h:
This file mirrors the kinetis common pin control file- is there a reason we can't include definitions from there?
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/pinctrl/pinctrl_soc_kinetis_common.h

danieldegrasse commented on dts/bindings/flash_controller/nxp,kinetis-ftfa.yaml:
The commit message here says:

> Some FTFA properties are configured in Kconfig. Add the properties in
dts binging as the properties configure harware.


Is this correct? I don't see where these properties are being moved from Kconfig to DTS. Perhaps the commit message needs to be something like the following?

> Add configuration settings for the FOPT, FSEC, and flash configuration offset to the FTFA module binding

michal-smola replied:
Commit message was changed as proposed.

michal-smola committed:
include: dt-bindings: kinetis sim: Add clock sources definitions

Add OSCERCLK, MCGIRCLK, and MCGPCLK clock sources definitions
to the kinetis_sim.h in include dt-bindings.

Signed-off-by: Michal Smola <michal.smola@nxp.com>

michal-smola committed:
dts: mcxc: Add initial support for NXP MCXC socs

Add initial support for NXP MCXC socs.

Signed-off-by: Michal Smola <michal.smola@nxp.com>

nordicjm commented on dts/bindings/clock/nxp,kinetis-mcg.yaml:
```diff
@@ -14,5 +14,19 @@ properties:
   "#clock-cells":
     const: 1
 
+  fcrdiv:
+    type: int
+    description: |
+      Internal Reference Clock Divider.
+      Division factor is given as 2^fcrdiv.
+      The valid range is 0 to 7.
+
+  lircdiv2:
+    type: int
+    description: |
+      Second Low-frequency Internal Reference Clock Divider.
+      Division factor is given as 2^lircdiv2.
+      The valid range is 0 to 7.
```
why can't these be enums?

michal-smola replied:
@nordicjm, thank you for review. Changed to enum.

nordicjm commented on dts/bindings/flash_controller/nxp,kinetis-ftfa.yaml:
```diff
@@ -3,3 +3,22 @@ description: NXP Kinetis Flash Memory Module A (FTFA)
+    description: |
+      Configures the reset value of the FSEC register, which includes
+      backdoor key access, mass erase, factory access, and flash security
+      options.
```
seems like these could be enums too?

michal-smola replied:
Prefered to be int.

nordicjm commented on boards/nxp/frdm_mcxc242/doc/index.rst:
```diff
@@ -0,0 +1,221 @@
+Supported Features
+==================
+
+The FRDM-MCXC242 board configuration supports the following hardware features:
```
use west board name with \`\` around it and *board target

michal-smola replied:
Fixed.

nordicjm commented on boards/nxp/frdm_mcxc242/frdm_mcxc242_defconfig:
```diff
@@ -0,0 +1,11 @@
+CONFIG_UART_CONSOLE=y
+CONFIG_SERIAL=y
+CONFIG_GPIO=y
+CONFIG_PINCTRL=y
```
no pinctrl selection at board level

michal-smola replied:
Removed pinctrl selection

nordicjm reviewed (changes requested):
last 2 comments are blockers, first 2 are optional
